File: Data Structures & Algorithms/jump-game/submission-1.cpp
class Solution {
public:
    bool solve(vector<int>& nums,int index,auto& dp){
        if(index >= nums.size()-1)return true;
        if(dp[index] != -1)return dp[index];
        for(int i=1;i<=nums[index];i++){
            if(solve(nums,index+i,dp)){
                return dp[index] = true;
            }
        }
        return dp[index]= false;
    }
    bool canJump(vector<int>& nums) {
        vector<int> dp(nums.size(),-1);
        return solve(nums,0,dp);
    }
};

File: Data Structures & Algorithms/subsets/submission-1.java
class Solution {
    public void solve(List<List<Integer>> ans,List<Integer> temp,int index,int[] nums){
        if(index == nums.length){
            ans.add(new ArrayList<>(temp));
            return;
        }
        // not take
        solve(ans,temp,index+1,nums);
        temp.add(nums[index]);
        solve(ans,temp,index+1,nums);
        temp.remove(temp.size()-1);
    }
    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> ans = new ArrayList<>();
        List<Integer> temp = new ArrayList<>();
        solve(ans,temp,0,nums);
        return ans;
    }
}
